src/streamer.ts

```diff
@@ -0,0 +1,36 @@
+import { S3Client } from '@aws-sdk/client-s3';
+import { listBlocks, fetchStreamerMessage } from './s3fetchers';
+import { LakeConfig, BlockHeight, StreamerMessage } from './types';
+
+export async function startStream(
+  config: LakeConfig,
+  onStreamerMessageReceived: (data: StreamerMessage) => Promise<void>
+) {
+  const s3Client = new S3Client({ region: config.s3RegionName });
+
+  let lastProcessedBlockHash: string;
+  let startFromBlockHeight = config.startBlockHeight;
+  let queue: Promise<void>[] = [];
+
+  while (true) {
+    const blockHeights = await listBlocks(s3Client, config.s3BucketName, startFromBlockHeight);
+    if (blockHeights.length > 0) {
+      startFromBlockHeight = blockHeights[blockHeights.length - 1];
+    } else {
+      setTimeout(() => {}, 2000);
+      continue;
+    }
+    for (let blockHeight of blockHeights) {
+      const streamerMessage = await fetchStreamerMessage(s3Client, config.s3BucketName, blockHeight);
+      if (lastProcessedBlockHash && lastProcessedBlockHash !== streamerMessage.block.header.prevHash) {
+        startFromBlockHeight = blockHeight - 1;
```

Comment:
In what case is this case needed?

Reply:
In case the instance of indexer built on top of this library follows the latest block. The nodes that write the data to S3 might store block 2 earlier than block 1. These lines ensure we don't stream block 2 to the indexer before block 1.